### ISSUE TITLE
Often get autocomplete 100-300% on macOS

### ISSUE THREAD

**Issue by taybart** [fish-shell/fish-shell#5949]
```
$ fish --version                                                                                                                                                                             
fish, version 3.0.2

$ uname -a
Darwin grove.local 18.6.0 Darwin Kernel Version 18.6.0: Thu Apr 25 23:16:27 PDT 2019; root:xnu-4903.261.4~2/RELEASE_X86_64 x86_64

$ echo $TERM
xterm-256color
```

No customization
I often get stuck/frozen 100-300% processes when using autocomplete for `ls`. This tends to be inside of git repositories, but is not always the case. Let me know if I can help/if you need more information.

**Comment by ridiculousfish** [comment]
Can you please use Activity Monitor or `sample` to take a  sample of the runaway fish, and share it? Thanks!

**Comment by taybart** [comment]
[Sample of fish.txt](https://github.com/fish-shell/fish-shell/files/3315333/Sample.of.fish.txt)

**Comment by taybart** [comment]
<img width="181" alt="image" src="https://user-images.githubusercontent.com/3513897/59937590-1a1ce680-9410-11e9-99b3-1b307ac4e529.png">
<img width="449" alt="image" src="https://user-images.githubusercontent.com/3513897/59937602-21dc8b00-9410-11e9-8eb7-700f6752c1d5.png">

**Comment by ridiculousfish** [comment]
Thanks. This looks like you have a cycle in your completion wrapping: A wraps B wraps A... We don't attempt to detect this yet.

Once fish has entered this 100% CPU and then (hopefully) recovered, what does `complete | grep wraps` show?

**Comment by taybart** [comment]
I have not gotten one to complete yet (although I have only waited ~5 min). After playing around a bit, it seems to be isolated to inside of tmux sessions and I am able to make it happen every time I try inside them. I will leave one for a bit longer this time, I don't know if it will complete ever...

**Comment by taybart** [comment]
Waited ~1 hour going to call it, TOD now, seems to be an infinite loop. Is there any special reason that having tmux as a parent process would cause wraps to not work correctly? This is also isolated to `ls` and not other tab completions.

**Comment by faho** [comment]
@taybart: Do you have any aliases or functions or custom completions, especially anything related to `ls`?

Do you have any differences in environment when running tmux? See `set`, especially the values of $fish_complete_path and $fish_function_path.

What will cause this is something like

```fish
complete -c ls -w foo # complete ls like foo
complete -c foo -w ls # complete foo like ls
```

and it appears that you have that with tmux, but not without.

**Comment by taybart** [comment]
Ah, I think I got it. One thing I like to do is have a default alias for instance `alias ls="ls --color -l -h"` and then source a local file that I keep for different locations (for mac `alias ls="ls -G -l -h"`). Removing the "default" alias was able to stop the autocomplete. Weirdly there were no issues with this double alias except for the autocomplete.

**Comment by faho** [comment]
> Weirdly there were no issues with this double alias except for the autocomplete.

It wasn't actually a double alias anymore - fish doesn't really have aliases, `alias` is just a function that defines functions, and if it detects a nameclash, it'll prefix the call with `command` or `builtin` as appropriate.

So your alias would end up like

```fish
function ls --wraps "ls -G -l -h" # the "wraps" being for completions
    command ls -G -l -h
end
```

And then, when you try to complete it, fish will try to complete `ls` like `ls -G -l -h`. So it'll see how `ls` is completed, which it'll try to complete like `ls -G -l -h`. So it'll see how `ls` is completed, which it'll try to complete like `ls -G -l -h`. So it'll see how `ls` is completed, which it'll try to complete like `ls -G -l -h`. So it'll see how `ls` is completed, which it'll try to complete like `ls -G -l -h`. So it'll see how `ls` is completed, which it'll try to complete like `ls -G -l -h`. So it'll see how `ls` is completed, which it'll try to complete like `ls -G -l -h`.

Yeah, this requires cycle detection.

**Comment by faho** [comment]
>One thing I like to do is have a default alias for instance alias ls="ls --color -l -h"

Though I have to chastise you a little bit, @taybart. An alias very much counts as "customization". Multiple aliases doubly so.

**Comment by taybart** [comment]
@faho you're right! My bad. Thanks for your work here! Much appreciated!

**Comment by ridiculousfish** [comment]
Let's leave this open, we can probably improve this situation.

**Comment by bokysan** [comment]
Hello,

I'm also running into a same / similar situation.

```
$ fish --version
fish, version 3.0.2
 $ uname -a
Darwin shakira.local 18.7.0 Darwin Kernel Version 18.7.0: Thu Jun 20 18:42:21 PDT 2019; root:xnu-4903.270.47~4/RELEASE_X86_64 x86_64
$ echo $TERM
xterm-256color
```

When using `cd` or `ls` fish would sometimes go into an infinite loop spending all available CPU. This wouldn't stop even after a few hours.

I can't pinpoint exact cases when this happens, but it's always when autocompleting files (e.g. fish would show part of file and when i click `<TAB>` to complete it, `fish` would block).

My `complete` is as follows:
```
complete --exclusive --command workon --arguments '(vf ls)'
complete --exclusive --command rmvirtualenv --arguments '(vf ls)'
complete --exclusive --command vf --arguments '(vf ls)' --condition '__vfcompletion_using_command rm'
complete --exclusive --command vf --arguments '(vf ls)' --condition '__vfcompletion_using_command activate'
complete --exclusive --command vf --description '# Defined in /usr/local/lib/python3.7/site-packages/virtualfish/virtual.fish @ line 190' --arguments tmp --condition __vfcompletion_needs_command
complete --exclusive --command vf --description '# Defined in /usr/local/lib/python3.7/site-packages/virtualfish/virtual.fish @ line 156' --arguments rm --condition __vfcompletion_needs_command
complete --exclusive --command vf --description '# Defined in /usr/local/lib/python3.7/site-packages/virtualfish/virtual.fish @ line 124' --arguments new --condition __vfcompletion_needs_command
complete --exclusive --command vf --description '# Defined in /usr/local/lib/python3.7/site-packages/virtualfish/virtual.fish @ line 169' --arguments ls --condition __vfcompletion_needs_command
complete --exclusive --command vf --description '# Defined in /usr/local/lib/python3.7/site-packages/virtualfish/virtual.fish @ line 283' --arguments help --condition __vfcompletion_needs_command
complete --exclusive --command vf --description '# Defined in /usr/local/lib/python3.7/site-packages/virtualfish/virtual.fish @ line 304' --arguments globalpackages --condition __vfcompletion_needs_command
complete --exclusive --command vf --description '# Defined in /usr/local/lib/python3.7/site-packages/virtualfish/virtual.fish @ line 85' --arguments deactivate --condition __vfcompletion_needs_command
complete --exclusive --command vf --description '# Defined in /usr/local/lib/python3.7/site-packages/virtualfish/virtual.fish @ line 275' --arguments connect --condition __vfcompletion_needs_command
complete --exclusive --command vf --description '# Defined in /usr/local/lib/python3.7/site-packages/virtualfish/virtual.fish @ line 185' --arguments cdpackages --condition __vfcompletion_needs_command
complete --exclusive --command vf --description '# Defined in /usr/local/lib/python3.7/site-packages/virtualfish/virtual.fish @ line 177' --arguments cd --condition __vfcompletion_needs_command
complete --exclusive --command vf --description '# Defined in /usr/local/lib/python3.7/site-packages/virtualfish/virtual.fish @ line 241' --arguments all --condition __vfcompletion_needs_command
complete --exclusive --command vf --description '# Defined in /usr/local/lib/python3.7/site-packages/virtualfish/virtual.fish @ line 204' --arguments addpath --condition __vfcompletion_needs_command
complete --exclusive --command vf --description '# Defined in /usr/local/lib/python3.7/site-packages/virtualfish/virtual.fish @ line 44' --arguments activate --condition __vfcompletion_needs_command
complete --no-files --command serverless --description serverless --arguments '(eval _serverless_completion)'
complete --no-files --command sls --description sls --arguments '(eval _sls_completion)'
complete --command \'\[\'
complete --command cat --short-option 'v' --description Display\ non-printing\ characters\ so\ they\'re\ visible.
complete --command cat --short-option 'u' --description 'Disable output buffering'
complete --command cat --short-option 't' --description Display\ non-printing\ characters,\ and\ tab\ characters\ as\ `\^I\'
complete --command cat --short-option 's' --description 'Single spaced output by squeezing adjacent empty lines'
complete --command cat --short-option 'n' --description 'Number all lines'
complete --command cat --short-option 'e' --description Display\ non-printing\ characters,\ and\ `\$\'\ at\ the\ end\ of\ each\ line
complete --command cat --short-option 'b' --description 'Number non-blank lines'
complete --command  --wraps
complete --command  --wraps
complete --command  --wraps
complete --command  --wraps
complete --command [ --wraps test
complete --command ll --wraps ls -lah
complete --command ls --wraps ls --color
complete --command ls --wraps ls -G
```

Any idea on how to track down what's causing it would be greatly appreciated.

**Comment by zanchey** [comment]
@bokysan please see the suggestions in https://github.com/fish-shell/fish-shell/issues/5949#issuecomment-504486727

**Comment by bokysan** [comment]
Please find attached. Killed on the following command after pressing TAB:
![image](https://user-images.githubusercontent.com/4567693/64942679-faa82500-d869-11e9-84fc-caca31343bb4.png)

[2019-09-16 Sample of fish.txt](https://github.com/fish-shell/fish-shell/files/3615217/2019-09-16.Sample.of.fish.txt)

**Comment by ridiculousfish** [comment]
@bokysan  you have a "wraps loop" - same as #5638. Did you define an alias for `ls`?

I should try to fix #5638, it's just going to keep biting people.

**Comment by ridiculousfish** [comment]
@bokysan I think this should be fixed in master, if you are able to test master can you please try on top of 5f64972908949b423170482b058446630d350571  ? Thank you for filing either way!
